File: example_cpp/draw2d.cpp
// Drawing a OBJ 2d-polygon
#include <cmath>
#include <ctime>
#include <GLFW/glfw3.h>
#include <GL/gl.h>
//#include <GL/glu.h>
#include <GLM/glm.hpp>
#include <glm/gtc/type_ptr.hpp>
#include "FileObj.h"
 
// Global variables
const char* title= "Polygon"; // Windowed mode's title
int windowWidth  = 600; // Windowed mode's width
int windowHeight = 600; // Windowed mode's height
int windowPosX   = 50;  // Windowed mode's top-left corner x
int windowPosY   = 50;  // Windowed mode's top-left corner y
 
// Projection clipping area
// clipAreaXLeft, clipAreaXRight, clipAreaYBottom, clipAreaYTop;
FileObj state;
GLuint  stateGL[] = { GLuint(-1), GLuint(-1), GLuint(-1) };


bool checkErrorGL () {
    GLenum err = glGetError(); 
    assert(err == GL_NO_ERROR);
    return (err != GL_NO_ERROR);
}

// Initialize OpenGL Graphics 
void initGL() {
   glClearColor(0.0, 0.0, 0.0, 1.0); // Set background (clear) color to black
   glClearDepth(1.0f);
   //glEnable(GL_DEPTH_TEST);
}
void releaseGL() {
    if (stateGL[0] != GLuint(-1)) {
        glDeleteLists(stateGL[0], 1);
        stateGL[0] = GLuint(-1);
    }
}
// Callback handler for window re-paint event
void display(GLFWwindow* window) 
{
  float center_x = 0.5*(state.bbox_min[0] + state.bbox_max[0]);
  float center_y = 0.5*(state.bbox_min[1] + state.bbox_max[1]);
  std::cout << "model-center" << center_x << ", " << center_y << std::endl;
  glClear  (GL_COLOR_BUFFER_BIT); // Clear the color buffer
  //glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT); // Clear the color and depth buffer
  glMatrixMode(GL_MODELVIEW); // To operate on the model-view matrix
  glLoadIdentity();           // Reset model-view matrix
  glTranslated  (-center_x, -center_y, 0.0);
	
  // Use triangular segments to form a circle
  glLineWidth(2.0f);
  glBegin(GL_LINE_LOOP);
  glColor3f  (1.0f, 0.0f, 0.0f);  // Red
  auto poly = state.getPolygon();
  for (auto iter = poly.begin(); iter != poly.end(); ++iter) { // Last vertex same as first vertex
      glVertex2f(iter->x, iter->y);
  }
  glEnd();
  checkErrorGL();
  glfwSwapBuffers(window);
}
// Callback handler for window re-paint event, using display lists
void displayDisplayList(GLFWwindow* window)
{
  float center_x = 0.5*(state.bbox_min[0] + state.bbox_max[0]);
  float center_y = 0.5*(state.bbox_min[1] + state.bbox_max[1]);
  std::cout << "model-center" << center_x << ", " << center_y << std::endl;
  glClear     (GL_COLOR_BUFFER_BIT); // Clear the color buffer
  glMatrixMode(GL_MODELVIEW); // To operate on the model-view matrix
  glLoadIdentity();           // Reset model-view matrix
  glTranslated  (-center_x, -center_y, 0.0);

  if (stateGL[0] == GLuint(-1)) {
      stateGL[0] = glGenLists(1);
  }
  glNewList(stateGL[0], GL_COMPILE_AND_EXECUTE); // Use triangular segments to form a circle
    checkErrorGL();
    glLineWidth(2.0);
    glBegin(GL_LINE_LOOP);
    glColor3f  (1.0, 0.0, 0.0); // Red
    auto poly = state.getPolygon();
    for (auto iter = poly.begin(); iter != poly.end(); ++iter) { // Last vertex same as first vertex
	glVertex2fv(*iter);
    }
    glEnd();
  glEndList();
  checkErrorGL();
  glfwSwapBuffers(window);
}
// Call back when the windows is re-sized */
void resize(GLFWwindow* window, int width, int height) 
{
    // Compute aspect ratio of the new window
    if (height == 0) {
        height = 1; // To prevent divide by 0
    }
        float aspect = float(width) / float(height);
        // Set the viewport to cover the new window
        glViewport(0, 0, width, height);

        // Set the aspect ratio of the clipping area to match the viewport
	    float size_x = 0.5f*(state.bbox_max[0] - state.bbox_min[0]); // refactor: move to "state"
	    float size_y = 0.5f*(state.bbox_max[1] - state.bbox_min[1]);
	    float size   = std::max(size_x, size_y);
        glMatrixMode(GL_PROJECTION); // To operate on the Projection matrix
        glLoadIdentity();            // Reset the projection matrix
        float clipAreaXLeft, clipAreaXRight, clipAreaYBottom, clipAreaYTop;
        if (width >= height) {
            clipAreaXLeft   = -size * aspect; 
            clipAreaXRight  =  size * aspect;
            clipAreaYBottom = -size;
            clipAreaYTop    =  size;
        }
        else {
            clipAreaXLeft   = -size;
            clipAreaXRight  =  size;
            clipAreaYBottom = -size / aspect; 
            clipAreaYTop    =  size / aspect;
        }
        glm::mat4 projection = glm::ortho(clipAreaXLeft, clipAreaXRight, clipAreaYBottom, clipAreaYTop);
        glLoadMatrixf(glm::value_ptr(projection));
        //gluOrtho2D(clipAreaXLeft, clipAreaXRight, clipAreaYBottom, clipAreaYTop);
}
 
// Called back when the timer expired 
//def Timer(value : int) -> None:
//glutTimerFunc(10, Timer, 0) // subsequent timer call at milliseconds
 
int main(int argc, char* argv[]) 
{
  //state.read("../star.obj");
  state.read("../nrw.obj");

  // Initialize glfw library (window toolkit).
  if (!glfwInit()) {
		return -1;
  } 

	// GLFW: Create a window and opengl context (version 4.3 compatibility profile).
	glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR, 4);
	glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, 3);
	glfwWindowHint(GLFW_OPENGL_PROFILE, GLFW_OPENGL_COMPAT_PROFILE);
    // OpenGL 4.3 core profile uses GLSL shaders
    //glfwWindowHint(GLFW_OPENGL_PROFILE, GLFW_OPENGL_CORE_PROFILE);
    glfwWindowHint(GLFW_RESIZABLE, GL_TRUE);
    glfwWindowHint(GLFW_DOUBLEBUFFER, GL_TRUE);
    glfwWindowHint(GLFW_DEPTH_BITS, 24);
    glfwWindowHint(GLFW_RESIZABLE, GL_TRUE);
    glfwWindowHint(GLFW_DOUBLEBUFFER, GL_TRUE);
	// Turn this on for smoother lines.
	// glfwWindowHint(GLFW_SAMPLES, 8);
	
    GLFWwindow* window = glfwCreateWindow(windowWidth, windowHeight, title, nullptr, nullptr);
	if (!window) {
		glfwTerminate();
		return -2;
	}

	// Make the window's opengl context current.
	glfwMakeContextCurrent(window);

    // Set callbacks for resize and keyboard events.
	glfwSetWindowSizeCallback   (window, resize);
    glfwSetWindowRefreshCallback(window, displayDisplayList);
	//fwSetKeyCallback(window, keyboard);

    // own OpenGL initialization
    initGL();                      

    // We have to call resize once for a proper setup.
	resize(window, windowWidth, windowHeight);

	// Loop until the user closes the window.
	while (!glfwWindowShouldClose(window)) {
        //glfwPollEvents();           // poll for and process events and change state (busy idle processing)
		glfwWaitEvents();           // wait for process events and change state (no idle processing)
        //glfwWaitEventsTimeout(0.033);  // sleep for up to 33ms to reduce CPU usage (idle processing)
        displayDisplayList(window); // then display
	}

	// Clean up everything on termination.
	releaseGL();
	glfwTerminate();

	return 0;

}

File: example_PyOpenGL/draw2d.py
'''
 Drawing a OBJ 2d-polygon
'''
from dataclasses import dataclass

from OpenGL.GL import *
from OpenGL.GLU import *
from OpenGL.GLUT import *
from math import *
import FileObj
import random
 
# Global variables
title        = b"Polygon" # Windowed mode's title
windowWidth  = 600 # Windowed mode's width
windowHeight = 600 # Windowed mode's height
windowPosX   = 50  # Windowed mode's top-left corner x
windowPosY   = 50  # Windowed mode's top-left corner y
 
# Projection clipping area
# clipAreaXLeft, clipAreaXRight, clipAreaYBottom, clipAreaYTop;
@dataclass
class AppState:
	polygon  = FileObj.FileObj()
	state_gl = [ -1 ]

state = AppState()

# Initialize OpenGL Graphics 
def initGL():
	glClearColor(1.0, 1.0, 1.0, 1.0) # Set background (clear) color to black

	# for polygonal types, not line types
	glFrontFace  (GL_CCW) # this is the default 
	glPolygonMode(GL_FRONT, GL_FILL) # front face (Vorderseite)
	glPolygonMode(GL_BACK,  GL_LINE) # back  face (Rückseite)

def drawGeometry():
	# Outline polygon
	glLineWidth(2.0)
	glBegin(GL_POLYGON)
	#glBegin(GL_TRIANGLE_FAN)
	#glBegin(GL_LINE_LOOP)
	point = state.polygon.getPointCoords()
	glColor3f  (1.0, 0.0, 0.0)  # Red
	for face in state.polygon.getPolygonIndices():
		for i, idx in enumerate(face): # Last vertex same as first vertex
			#glColor3f  (random.random(), random.random(), random.random())  # Red
			glVertex2fv(point[idx])
	glEnd()

# Callback handler for window re-paint event
def display():
	glClear     (GL_COLOR_BUFFER_BIT) # Clear the color buffer
	center_x = 0.5*(state.polygon.x[0] + state.polygon.x[1]) # refactor: move to "state"
	center_y = 0.5*(state.polygon.y[0] + state.polygon.y[1])
	print("model-center", center_x, center_y)
	glMatrixMode(GL_MODELVIEW)  # To operate on the ModelView matrix
	glLoadIdentity()
	glTranslated  (-center_x, -center_y, 0.0)

	drawGeometry()
	glutSwapBuffers()  # Swap front and back buffers (of double buffered mode)

# Callback handler for window re-paint event, using display lists
def displayDisplayList():
	glClear     (GL_COLOR_BUFFER_BIT) # Clear the color buffer
	center_x = 0.5*(state.polygon.x[1] + state.polygon.x[0]) # refactor: move to "state"
	center_y = 0.5*(state.polygon.y[1] + state.polygon.y[0])
	print("model-center", center_x, center_y)
	glMatrixMode(GL_MODELVIEW)  # To operate on the ModelView matrix
	glLoadIdentity()
	glTranslated  (-center_x, -center_y, 0.0)

	if state.state_gl[0] < 0: # not compiled, then compile and execute
		glNewList(state.state_gl[0], GL_COMPILE_AND_EXECUTE)
		drawGeometry()
		glEndList()
	else: # execute display-list
		glCallList(state.state_gl[0])
	glutSwapBuffers()  # Swap front and back buffers (of double buffered mode)
 
# Call back when the windows is re-sized */
def reshape(width, height):
	# Compute aspect ratio of the new window
	if height == 0: 
		height = 1 # To prevent divide by 0
	aspect = width / float(height)
	# Set the viewport to cover the new window
	glViewport(0, 0, width, height)
 
	# Set the aspect ratio of the clipping area to match the viewport
	glMatrixMode(GL_PROJECTION)  # To operate on the Projection matrix
	glLoadIdentity()             # Reset the projection matrix
	size_x = 0.5*(state.polygon.x[1] - state.polygon.x[0]) # refactor: move to "state"
	size_y = 0.5*(state.polygon.y[1] - state.polygon.y[0])
	size   = max(size_x, size_y)
	if width >= height:
		clipAreaXLeft   = -size * aspect
		clipAreaXRight  =  size * aspect
		clipAreaYBottom = -size
		clipAreaYTop    =  size
	else:
		clipAreaXLeft   = -size
		clipAreaXRight  =  size
		clipAreaYBottom = -size / aspect
		clipAreaYTop    =  size / aspect
	gluOrtho2D(clipAreaXLeft, clipAreaXRight, clipAreaYBottom, clipAreaYTop)
 
# Called back when the timer expired 
#def Timer(value : int) -> None:
#glutTimerFunc(10, Timer, 0) # subsequent timer call at milliseconds
 
# Main function: GLUT runs as a console application starting at main() */
def main():
	state.polygon.read("../models/5gon.obj")
	#state.polygon.read("../models/nrw.obj")

	glutInit(sys.argv)             # Initialize GLUT
	glutInitDisplayMode(GLUT_DOUBLE) # Enable double buffered mode
	glutInitWindowSize (windowWidth, windowHeight)  # Initial window width and height
	glutInitWindowPosition(windowPosX, windowPosY)  # Initial window top-left corner (x, y)
	glutCreateWindow(title)       # Create window with given title
	glutReshapeFunc(reshape)      # Register callback handler for window re-shape
	glutDisplayFunc(display)      # Register callback handler for window re-paint
	#glutIdleFunc(display)         # Register callback handler for window idling: redisplay
	#glutTimerFunc(0, Timer, 0)    # First timer call immediately
	
	initGL()                      # Our own OpenGL initialization
	glutMainLoop()                # Enter event-processing loop

if __name__ == '__main__':
	main()
